File: logger2.cpp
#include <iostream>
#include <mutex>
#include <thread>
#include <vector>

int main() {
  std::vector<int> logger;

  std::vector<std::thread> handlers;
  std::mutex mtx;
  for(int i = 0; i<10; i++) {
    handlers.push_back(std::thread([i, &logger, &mtx] {
      std::lock_guard<std::mutex> lg(mtx);
      logger.push_back(i);
    }));
  }

  for(auto& h : handlers) {
    h.join();
  }

  for(auto log : logger) {
    std::cout << log << ", ";
  }
  std::cout << std::endl;

  return 0;
}

File: logger2.rs
use std::sync::Arc;
use std::thread;

fn main() {
    let mut logger = Arc::new(vec![]);

    let mut handlers = vec![];
    for i in 0..10 {
        let logger_cloned = logger.clone();
        handlers.push(
            thread::spawn(move || {
                // ... do something ...

                // XXX cannot compile because here:
                logger_cloned.push(i);
            }
        ));
    }

    for h in handlers {
        h.join().unwrap();
    }
}
